<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
  </style>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <!-- jQuery -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/jqueryui/1.10.3/jquery-ui.min.js"></script>
  
  <!-- bootstrap -->
  <!--<link href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" rel="stylesheet"  id="style">-->
  <script src="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/js/bootstrap.min.js"></script>
  
  <!-- highlight.js -->
  <!--<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" rel="stylesheet" id="code-style">-->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/highlight.min.js"></script>
  <script>
  hljs.LANGUAGES.r=function(a){var b="([a-zA-Z]|\\.[a-zA-Z.])[a-zA-Z0-9._]*";return{c:[a.HCM,{b:b,l:b,k:{keyword:"function if in break next repeat else for return switch while try tryCatch|10 stop warning require library attach detach source setMethod setGeneric setGroupGeneric setClass ...|10",literal:"NULL NA TRUE FALSE T F Inf NaN NA_integer_|10 NA_real_|10 NA_character_|10 NA_complex_|10"},r:0},{cN:"number",b:"0[xX][0-9a-fA-F]+[Li]?\\b",r:0},{cN:"number",b:"\\d+(?:[eE][+\\-]?\\d*)?L\\b",r:0},{cN:"number",b:"\\d+\\.(?!\\d)(?:i\\b)?",r:0},{cN:"number",b:"\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",r:0},{b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[a.BE],r:0},{cN:"string",b:"'",e:"'",c:[a.BE],r:0}]}}(hljs); </script>
  <!--<script type="text/javascript", src="https://yandex.st/highlightjs/7.3/languages/r.min.js"></script>-->
  
  <!-- Manific Popup -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/magnific-popup.js/0.8.9/jquery.magnific-popup.min.js"></script>
  
  <script type="text/javascript"
    src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
  </script>
  <script defer="defer">
  // Function to generate the dynamic table of contents
  jQuery.fn.generate_TOC = function () {
    var base = $(this[0]);
  
    var selectors = ['h1', 'h2', 'h3', 'h4'];
  
    var last_ptr = [{}, {}, {}, {}];
  
    var anchors = {};
  
    generate_anchor = function (text) {
      var test = text.replace(/\W/g, '_');
  
      while(test in anchors){
        //if no suffix, add one
        if(test.match(/_\d+$/) === null){
          test = test + "_2";
        }
        //else generate unique id for duplicates by adding one to the suffix
        else {
          test = test.replace(/_(\d+)$/, function(match, number){ var num=+number+1; return("_" + num) });
        }
      }
      anchors[test]=1;
      return(test);
    }
  
    $(selectors.join(',')).filter(function(index) { return $(this).parent().attr("id") != 'header'; }).each(function () {
  
      var heading = $(this);
      var idx = selectors.indexOf(heading.prop('tagName').toLowerCase());
      var itr = 0;
  
      while (itr <= idx) {
        if (jQuery.isEmptyObject(last_ptr[itr])) {
          last_ptr[itr] = $('<ul>').addClass('nav');
          if (itr === 0) {
            base.append(last_ptr[itr])
          } else {
            if(last_ptr[itr-1].children('li').length === 0){
              last_ptr[itr-1].append(last_ptr[itr]);
            }
            else {
              last_ptr[itr - 1].children('li').last().append(last_ptr[itr]);
            }
          }
        }
        itr++;
      }
      var anchor = generate_anchor(heading.text());
      heading.attr('id', anchor);
      var a = $('<a>')
      .text(heading.text())
      .attr('href', '#' + anchor);
  
    var li = $('<li>')
      .append(a);
  
    last_ptr[idx].append(li);
    for (i = idx + 1; i < last_ptr.length; i++) {
      last_ptr[i] = {};
    }
    });
  }
  /* run scripts when document is ready */
  $(function() {
    "use strict";
  
    var $window = $(window);
    var $body = $(document.body);
  
    /* size of thumbnails */
  
    var hidden_types = ['source']
    var output_types = ['output', 'message', 'warning', 'error']
  
    /* style tables */
    $('table').addClass('table table-striped table-bordered table-hover table-condensed');
  
    $('pre code').each(function(i, e) {
      hljs.highlightBlock(e);
    });
  
    /* Magnific Popup */
    $(".thumbnail").each(function(){
      $(this).magnificPopup({
        disableOn: 768,
        closeOnContentClick: true,
  
        type: 'image',
        items: {
          src: $(this).find('img').attr('src'),
        }
      });
    });
  
    function toggle_block(obj, show) {
      var span = obj.find('span');
      if(show === true){
        span.removeClass('glyphicon-chevron-up').addClass('glyphicon-chevron-down');
        obj.next('pre').slideDown();
      }
      else {
        span.removeClass('glyphicon-chevron-down').addClass('glyphicon-chevron-up');
        obj.next('pre').slideUp();
      }
    }
  
    function toggle_thumbnails(imgs, show){
      if(show === true){
        imgs.parents().show()
        imgs.slideDown();
      }
      else {
        imgs.slideUp(400, function(){ $(this).parent().hide(); });
      }
    }
  
    function global_toggle(obj){
      var type = obj.attr('type');
      var show = !obj.parent('li').hasClass('active');
      if(show === true){
        obj.parent('li').addClass('active');
      }
      else{
        obj.parent('li').removeClass('active');
      }
      if(type == 'figure'){
        toggle_thumbnails($('.thumbnail img'), show);
      }
      else {
        $('.toggle.' + type).each(function() { toggle_block($(this), show); });
      }
    }
  
    /* onclick toggle next code block */
    $('.toggle').click(function() {
      var span = $(this).find('span');
      toggle_block($(this), !span.hasClass('glyphicon-chevron-down'));
      return false
    })
  
    // global toggles
    $('.toggle-global').click(function(){
      var type = $(this).attr('type');
      if(type === 'all-source'){
          $('li a.source').each(function() {
            global_toggle($(this));
          });
        }
      else if(type === 'all-output'){
        $.each(output_types, function(i, val){
          console.log(val);
          global_toggle($('li a.' + val));
        });
      }
      else {
        console.log($(this));
        global_toggle($(this));
      }
      return false;
    });
    /* table of contents */
    if($(['h1', 'h2', 'h3', 'h4'].join(',')).length > 0){
      $('body > #wrap > .container > .row').append('<div class="col-md-2"><div id="toc" class="well sidebar sidenav affix hidden-print"/></div>');
      $('#toc').generate_TOC();
    }
  
    $.each(hidden_types, function(i, type) {
      $('li[type=' + type + ']').each(function(){ global_toggle($(this)); });
    });
  
    /* remove paragraphs with no content */
    $('p:empty').remove();
  
    $body.scrollspy({
      target: '.sidebar',
    });
  
    /* theme switch */
    $('.theme-switch').click(function(){
      var css = $('link[title=' + $(this).attr('title') + ']');
      $('#theme[rel=stylesheet]').attr('href', css.attr('href'));
      $('.theme-switch').closest('li').removeClass('active');
      $(this).closest('li').addClass('active');
      return false;
    });
    /* code style switch */ //TODO use same function for both of these?
    $('.highlight-switch').click(function(){
      var css = $('link[title="' + $(this).attr('title') + '"]');
      $('#highlight[rel=stylesheet]').attr('href', css.attr('href'));
      $('.highlight-switch').closest('li').removeClass('active');
      $(this).closest('li').addClass('active');
      return false;
    });
  
    //TODO refresh on show/hide
    $window.on('load', function () {
      $body.scrollspy('refresh');
    })
  
  });
  
  </script>
  <style>
  /* Knitr_bootstrap styles */
  #header {
    display: none !important;
    visibility: hidden !important;
  }
  #wrap .container-fluid {
    padding: 0;
    overflow: hidden;
  }
  .toggle{
    text-transform: capitalize;
  }
  
  .toggle-global{
    text-transform: capitalize;
  }
  
  /* Sticky footer styles */
  * {
    margin:0;
  }
  html,
  body {
      height: 100%;
      padding:0 !important;
      /* The html and body elements cannot have any padding or margin. */
      /*overflow-x: hidden;*/
  }
  
  /* Wrapper for page content to push down footer */
  #wrap {
      min-height: 100%;
      height: auto !important;
      height: 100%;
      /* Negative indent footer by it's height */
      margin: 0 auto -120px;
  }
  
  /* Set the fixed height of the footer here */
  #push,
  #footer {
      height: 120px;
  }
  
  #footer {
    text-align: center;
  }
  
  /* Top level subheader elements.  These are the first nested items underneath a header element. */
  .header li {
    font-size: 20px;
  }
  
  /* Makes the font smaller for all subheader elements. */
  .sub-header li {
      font-size: 12px;
  }
  
  button.thumbnails {
    margin-left:0px;
  }
  
  /*
   * Side navigation
   *
   * Scrollspy and affixed enhanced navigation to highlight sections and secondary
   * sections of docs content.
   */
  
  /* By default it's not affixed in mobile views, so undo that */
  .sidebar.affix {
    position: static;
  }
  
  /* First level of nav */
  .sidenav {
    margin-top: 30px;
    margin-bottom: 30px;
    padding-top:    10px;
    padding-bottom: 10px;
    border-radius: 5px;
  }
  
  /* All levels of nav */
  .sidebar .nav > li > a {
    display: block;
    padding: 5px 20px;
  }
  .sidebar .nav > li > a:hover,
  .sidebar .nav > li > a:focus {
    text-decoration: none;
    border-right: 1px solid;
  }
  .sidebar .nav > .active > a,
  .sidebar .nav > .active:hover > a,
  .sidebar .nav > .active:focus > a {
    font-weight: bold;
    background-color: transparent;
    border-right: 1px solid;
  }
  
  /* Nav: second level (shown on .active) */
  .sidebar .nav .nav {
    display: none; /* Hide by default, but at >768px, show it */
    margin-bottom: 8px;
  }
  .sidebar .nav .nav > li > a {
    padding-top:    3px;
    padding-bottom: 3px;
    padding-left: 30px;
    font-size: 90%;
  }
  
  .sidebar .nav .nav .nav > li > a {
    padding-left: 40px;
  }
  .sidebar .nav .nav .nav .nav > li > a {
    padding-left: 50px;
  }
  
  /* Show and affix the side nav when space allows it */
  @media screen and (min-width: 992px) {
    .sidebar .nav > .active > ul {
      display: block;
    }
    /* Widen the fixed sidebar */
    .sidebar.affix,
    .sidebar.affix-bottom {
      width: 213px;
    }
    .sidebar.affix-top,
    .sidebar.affix {
      position: fixed; /* Undo the static from mobile first approach */
      top: 30px;
    }
    .sidebar.affix-bottom {
      position: absolute; /* Undo the static from mobile first approach */
    }
    .sidebar.affix-bottom .sidenav,
    .sidebar.affix .sidenav {
      margin-top: 0;
      margin-bottom: 0;
    }
  }
  @media screen and (min-width: 1200px) {
    /* Widen the fixed sidebar again */
    .sidebar.affix-bottom,
    .sidebar.affix {
      width: 263px;
    }
  }
  
  #toc {
    padding: 10px 0px;
    margin:0;
    border:0;
  }
  
  
  .panel pre {
    margin: 0;
    padding: 0;
    border: 0;
  }
  button + pre {
    margin: 0;
    padding: 0;
  }
  pre code {
    border-radius: 0;
  }
  /* Magnific Popup CSS */
  .mfp-bg {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1042;
    overflow: hidden;
    position: fixed;
    background: #0b0b0b;
    opacity: 0.8;
    filter: alpha(opacity=80); }
  
  .mfp-wrap {
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 1043;
    position: fixed;
    outline: none !important;
    -webkit-backface-visibility: hidden; }
  
  .mfp-container {
    text-align: center;
    position: absolute;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    padding: 0 8px;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box; }
  
  .mfp-container:before {
    content: '';
    display: inline-block;
    height: 100%;
    vertical-align: middle; }
  
  .mfp-align-top .mfp-container:before {
    display: none; }
  
  .mfp-content {
    position: relative;
    display: inline-block;
    vertical-align: middle;
    margin: 0 auto;
    text-align: left;
    z-index: 1045; }
  
  .mfp-inline-holder .mfp-content,
  .mfp-ajax-holder .mfp-content {
    width: 100%;
    cursor: auto; }
  
  .mfp-ajax-cur {
    cursor: progress; }
  
  .mfp-zoom-out-cur,
  .mfp-zoom-out-cur .mfp-image-holder .mfp-close {
    cursor: -moz-zoom-out;
    cursor: -webkit-zoom-out;
    cursor: zoom-out; }
  
  .mfp-zoom {
    cursor: pointer;
    cursor: -webkit-zoom-in;
    cursor: -moz-zoom-in;
    cursor: zoom-in; }
  
  .mfp-auto-cursor .mfp-content {
    cursor: auto; }
  
  .mfp-close,
  .mfp-arrow,
  .mfp-preloader,
  .mfp-counter {
    -webkit-user-select: none;
    -moz-user-select: none;
    user-select: none; }
  
  .mfp-loading.mfp-figure {
    display: none; }
  
  .mfp-hide {
    display: none !important; }
  
  .mfp-preloader {
    color: #cccccc;
    position: absolute;
    top: 50%;
    width: auto;
    text-align: center;
    margin-top: -0.8em;
    left: 8px;
    right: 8px;
    z-index: 1044; }
  
  .mfp-preloader a {
    color: #cccccc; }
  
  .mfp-preloader a:hover {
    color: white; }
  
  .mfp-s-ready .mfp-preloader {
    display: none; }
  
  .mfp-s-error .mfp-content {
    display: none; }
  
  button.mfp-close,
  button.mfp-arrow {
    overflow: visible;
    cursor: pointer;
    background: transparent;
    border: 0;
    -webkit-appearance: none;
    display: block;
    padding: 0;
    z-index: 1046;
    -webkit-box-shadow: none;
    box-shadow: none; }
  
  button::-moz-focus-inner {
    padding: 0;
    border: 0; }
  
  .mfp-close {
    width: 44px;
    height: 44px;
    line-height: 44px;
    position: absolute;
    right: 0;
    top: 0;
    text-decoration: none;
    text-align: center;
    opacity: 0.65;
    padding: 0 0 18px 10px;
    color: white;
    font-style: normal;
    font-size: 28px;
    font-family: Arial, Baskerville, monospace; }
    .mfp-close:hover, .mfp-close:focus {
      opacity: 1; }
    .mfp-close:active {
      top: 1px; }
  
  .mfp-close-btn-in .mfp-close {
    color: #333333; }
  
  .mfp-image-holder .mfp-close,
  .mfp-iframe-holder .mfp-close {
    color: white;
    right: -6px;
    text-align: right;
    padding-right: 6px;
    width: 100%; }
  
  .mfp-counter {
    position: absolute;
    top: 0;
    right: 0;
    color: #cccccc;
    font-size: 12px;
    line-height: 18px; }
  
  .mfp-arrow {
    position: absolute;
    opacity: 0.65;
    margin: 0;
    top: 50%;
    margin-top: -55px;
    padding: 0;
    width: 90px;
    height: 110px;
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0); }
  
  .mfp-arrow:active {
    margin-top: -54px; }
  
  .mfp-arrow:hover,
  .mfp-arrow:focus {
    opacity: 1; }
  
  .mfp-arrow:before, .mfp-arrow:after,
  .mfp-arrow .mfp-b,
  .mfp-arrow .mfp-a {
    content: '';
    display: block;
    width: 0;
    height: 0;
    position: absolute;
    left: 0;
    top: 0;
    margin-top: 35px;
    margin-left: 35px;
    border: medium inset transparent; }
  .mfp-arrow:after,
  .mfp-arrow .mfp-a {
    border-top-width: 13px;
    border-bottom-width: 13px;
    top: 8px; }
  .mfp-arrow:before,
  .mfp-arrow .mfp-b {
    border-top-width: 21px;
    border-bottom-width: 21px; }
  
  .mfp-arrow-left {
    left: 0; }
    .mfp-arrow-left:after,
    .mfp-arrow-left .mfp-a {
      border-right: 17px solid white;
      margin-left: 31px; }
    .mfp-arrow-left:before,
    .mfp-arrow-left .mfp-b {
      margin-left: 25px;
      border-right: 27px solid #3f3f3f; }
  
  .mfp-arrow-right {
    right: 0; }
    .mfp-arrow-right:after,
    .mfp-arrow-right .mfp-a {
      border-left: 17px solid white;
      margin-left: 39px; }
    .mfp-arrow-right:before,
    .mfp-arrow-right .mfp-b {
      border-left: 27px solid #3f3f3f; }
  
  .mfp-iframe-holder {
    padding-top: 40px;
    padding-bottom: 40px; }
  
  .mfp-iframe-holder .mfp-content {
    line-height: 0;
    width: 100%;
    max-width: 900px; }
  
  .mfp-iframe-scaler {
    width: 100%;
    height: 0;
    overflow: hidden;
    padding-top: 56.25%; }
  
  .mfp-iframe-scaler iframe {
    position: absolute;
    display: block;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    box-shadow: 0 0 8px rgba(0, 0, 0, 0.6);
    background: black; }
  
  .mfp-iframe-holder .mfp-close {
    top: -40px; }
  
  /* Main image in popup */
  img.mfp-img {
    width: auto;
    max-width: 100%;
    height: auto;
    display: block;
    line-height: 0;
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    padding: 40px 0 40px;
    margin: 0 auto; }
  
  /* The shadow behind the image */
  .mfp-figure:after {
    content: '';
    position: absolute;
    left: 0;
    top: 40px;
    bottom: 40px;
    display: block;
    right: 0;
    width: auto;
    height: auto;
    z-index: -1;
    box-shadow: 0 0 8px rgba(0, 0, 0, 0.6);
    background: #444444; }
  
  .mfp-figure {
    line-height: 0; }
  
  .mfp-bottom-bar {
    margin-top: -36px;
    position: absolute;
    top: 100%;
    left: 0;
    width: 100%;
    cursor: auto; }
  
  .mfp-title {
    text-align: left;
    line-height: 18px;
    color: #f3f3f3;
    word-wrap: break-word;
    padding-right: 36px; }
  
  .mfp-figure small {
    color: #bdbdbd;
    display: block;
    font-size: 12px;
    line-height: 14px; }
  
  .mfp-image-holder .mfp-content {
    max-width: 100%; }
  
  .mfp-gallery .mfp-image-holder .mfp-figure {
    cursor: pointer; }
  
  @media screen and (max-width: 800px) and (orientation: landscape), screen and (max-height: 300px) {
    /**
     * Remove all paddings around the image on small screen
     */
    .mfp-img-mobile .mfp-image-holder {
      padding-left: 0;
      padding-right: 0; }
  
    .mfp-img-mobile img.mfp-img {
      padding: 0; }
  
    /* The shadow behind the image */
    .mfp-img-mobile .mfp-figure:after {
      top: 0;
      bottom: 0; }
  
    .mfp-img-mobile .mfp-bottom-bar {
      background: rgba(0, 0, 0, 0.6);
      bottom: 0;
      margin: 0;
      top: auto;
      padding: 3px 5px;
      position: fixed;
      -webkit-box-sizing: border-box;
      -moz-box-sizing: border-box;
      box-sizing: border-box; }
  
    .mfp-img-mobile .mfp-bottom-bar:empty {
      padding: 0; }
  
    .mfp-img-mobile .mfp-counter {
      right: 5px;
      top: 3px; }
  
    .mfp-img-mobile .mfp-close {
      top: 0;
      right: 0;
      width: 35px;
      height: 35px;
      line-height: 35px;
      background: rgba(0, 0, 0, 0.6);
      position: fixed;
      text-align: center;
      padding: 0; }
  
    .mfp-img-mobile .mfp-figure small {
      display: inline;
      margin-left: 5px; } }
  @media all and (max-width: 900px) {
    .mfp-arrow {
      -webkit-transform: scale(0.75);
      transform: scale(0.75); }
  
    .mfp-arrow-left {
      -webkit-transform-origin: 0;
      transform-origin: 0; }
  
    .mfp-arrow-right {
      -webkit-transform-origin: 100%;
      transform-origin: 100%; }
  
    .mfp-container {
      padding-left: 6px;
      padding-right: 6px; } }
  .mfp-ie7 .mfp-img {
    padding: 0; }
  .mfp-ie7 .mfp-bottom-bar {
    width: 600px;
    left: 50%;
    margin-left: -300px;
    margin-top: 5px;
    padding-bottom: 5px; }
  .mfp-ie7 .mfp-container {
    padding: 0; }
  .mfp-ie7 .mfp-content {
    padding-top: 44px; }
  .mfp-ie7 .mfp-close {
    top: 0;
    right: 0;
    padding-top: 0; }
  
  //Magnific overrides
  .mfp-image img{
    background: white;
  }
  .mfp-figure:after {
    background: white;
  }
  
  /*
   * Off Canvas navbar toggle right
   * --------------------------------------------------
   */
  
  @media screen and (max-width: 768px) {
    .row-offcanvas .collapsing {
    -webkit-transition: none 0;
      -moz-transition: none 0;
      transition: none 0;
    }
   .row-offcanvas .navbar {
    position: absolute;
    z-index: 2;
      right:0;
      height:100%;
      width:55px;
      border:0;
      background-color:transparent;
    }
    .row-offcanvas .navbar-toggle {
      margin-right: 5px;
      margin-left: 5px;
    }
    .row-offcanvas {
      position: relative;
    }
    .row-offcanvas-right.active .navbar {
    position: absolute;
    z-index: 2;
      right: -28.4%;
      width:40%;
      background-color:#eee;
      border:0 solid #ddd;
      border-left-width:1px;
    }
    .row-offcanvas-right.active {
      left: -30%;
    }
    .row-offcanvas-right.active .navbar-collapse {
      position: relative;
      width: 100%;
    }
    .row-offcanvas .content {
    /*width:calc(100% - 60px);*/
    }
  }
  </style>
</head>
<body>
<div id="wrap">
<div class="container">
<div class="row row-offcanvas row-offcanvas-right">
<div class="contents col-xs-12 col-md-10">
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{derfinder advanced details and usage}
-->
<h1><code>derfinder</code> advanced details and usage</h1>
<p>If you wish, you can view this vignette online <a href="http://lcolladotor.github.io/derfinder/">here</a>.</p>
<h1>Overview</h1>
<p>This vignette explains in more detail the relationship between the different functions in <code>derfinder</code> <a id='cite-derfinder'></a>(<a href='https://github.com/lcolladotor/derfinder'>Collado-Torres, Frazee, Jaffe, and Leek, 2014</a>) as well as add-on packages <code>derfinderPlot</code> (<a href="http://lcolladotor.github.io/derfinderPlot/">vignette</a>), <code>derfinderHelper</code> (<a href="http://lcolladotor.github.io/derfinderHelper/">vignette</a>) and <code>regionReport</code> (<a href="http://lcolladotor.github.io/regionReport/">vignette</a>). The vignette also includes some bash scripts for running <code>derfinder</code> in a cluster, although there are probably other ways to do so using only R scripts. For example, by using <code>BatchJobsParam()</code> from <a href="www.bioconductor.org/packages/release/bioc/html/BiocParallel.html">BiocParallel</a>.</p>
<p>This vignette assumes that you have read through the introductory vignette.</p>
<p>Lets start by loading the <code>derfinder</code> package.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Load libraries
library('derfinder')</code></pre>
</div>
<p>If you explore the package, you will notice that all exported functions use camel case name and have aliases with underscore names. For example, <code>analyzeChr()</code> and <code>analyze_chr()</code>. You should use whichever naming style you prefer. Regrettably, implementing the same flexibility at the argument naming level is not straightforward.</p>
<h1>Advanced arguments</h1>
<p>Just like it is shown in the introductory vignette, you can find more information about the advanced arguments using <code>advancedArg()</code>.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## URLs to advanced arguemtns
sapply(c('analyzeChr', 'loadCoverage'), advancedArg, browse = FALSE)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##                                                                                                                             analyzeChr 
## "https://github.com/search?utf8=%E2%9C%93&q=advanced_argument+filename%3Aanalyze+repo%3Alcolladotor%2Fderfinder+path%3A%2FR&type=Code" 
##                                                                                                                           loadCoverage 
##    "https://github.com/search?utf8=%E2%9C%93&q=advanced_argument+filename%3Aload+repo%3Alcolladotor%2Fderfinder+path%3A%2FR&type=Code"
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Set browse = TRUE if you want to open them in your browser</code></pre>
</div>
<p>These arguments are options we expect not all users will want to change and which would otherwise make the help pages confusing to them. However, all arguments are documented in <a href="http://cran.r-project.org/web/packages/roxygen2/index.html">roxygen2</a> style in the source code.</p>
<p>Furthermore, note that using the <code>...</code> argument allows you to specify some of the documented arguments. For example, you might want to control the <code>maxClusterGap</code> from <code>findRegions()</code> in the <code>analyzeChr()</code> call.</p>
<h1>Advanced loading data</h1>
<h2>Controlling loading from BAM files</h2>
<p>If you are loading data from BAM files, you might want to specify some criteria to decide which reads to include or not. For example, your data might have been generated by a strand-specific protocol. You can do so by specifying the arguments of <code>scanBamFlag()</code> from <a href="http://www.bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a>.</p>
<h2>Unfiltered base-level coverage</h2>
<p>Note that in most scenarios, the <code>fullCov</code> object illustrated in the introductory vignette can be large in memory. When making plots or calculating the region level coverage, we don't need the full information. In such situations, it might pay off to create a smaller version by loading only the required data. This can be achieved using the advanced argument <code>which</code> to <code>fullCoverage()</code> or <code>loadCoverage()</code>.</p>
<p>However, note that when reading the data from BAM files, a read might align partially inside the region of interest. By default such a read would be discarded and thus the base-level coverage would be lower than what it is in reality. The advanced argument <code>protectWhich</code> extends regions by 30 kbp (15 kbp each side) to help mitigate this issue.</p>
<p>We can illustrate this issue with the example data from <code>derfinder</code>. First, we load in the data and generate some regions of interest.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Find some regions to work with
example('loadCoverage', 'derfinder')</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-up"></span> R output
</button>
<pre style="display:none"><code class="output r">## 
## ldCvrg> datadir <- system.file('extdata', 'genomeData', package='derfinder')
## 
## ldCvrg> files <- rawFiles(datadir = datadir, samplepatt = '*accepted_hits.bam$',
## ldCvrg+     fileterm = NULL)
## 
## ldCvrg> ## Shorten the column names
## ldCvrg> names(files) <- gsub('_accepted_hits.bam', '', names(files))
## 
## ldCvrg> ## Read and filter the data, only for 2 files
## ldCvrg> dataSmall <- loadCoverage(files = files[1:2], chr = '21', cutoff = 0)
</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## 2014-09-30 14:50:44 loadCoverage: finding chromosome lengths
## 2014-09-30 14:50:44 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam
## 2014-09-30 14:50:44 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam
## 2014-09-30 14:50:44 loadCoverage: applying the cutoff to the merged data
## Loading required package: parallel
## 
## Attaching package: 'BiocGenerics'
## 
## The following objects are masked from 'package:parallel':
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB
## 
## The following object is masked from 'package:stats':
## 
##     xtabs
## 
## The following objects are masked from 'package:base':
## 
##     anyDuplicated, append, as.data.frame, as.vector, cbind,
##     colnames, do.call, duplicated, eval, evalq, Filter, Find, get,
##     intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rep.int, rownames, sapply, setdiff, sort,
##     table, tapply, union, unique, unlist, unsplit
## 
## Loading required package: stats4
## 2014-09-30 14:50:45 filterData: originally there were 48129895 rows, now there are 454 rows. Meaning that 100 percent was filtered.
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-up"></span> R output
</button>
<pre style="display:none"><code class="output r">## 
## ldCvrg> ## Not run: 
## ldCvrg> ##D ## Export to BigWig files
## ldCvrg> ##D createBw(list('chr21' = dataSmall))
## ldCvrg> ##D 
## ldCvrg> ##D ## Load data from BigWig files
## ldCvrg> ##D dataSmall.bw <- loadCoverage(c(ERR009101 = 'ERR009101.bw', ERR009102 =
## ldCvrg> ##D     'ERR009102.bw'), chr = 'chr21')
## ldCvrg> ##D 
## ldCvrg> ##D ## Compare them
## ldCvrg> ##D mapply(function(x, y) { x - y }, dataSmall$coverage, dataSmall.bw$coverage)
## ldCvrg> ##D 
## ldCvrg> ##D ## Note that the only difference is the type of Rle (integer vs numeric) used
## ldCvrg> ##D ## to store the data.
## ldCvrg> ##D 
## ldCvrg> ## End(Not run)
## ldCvrg> 
## ldCvrg> 
## ldCvrg>
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">example('getRegionCoverage', 'derfinder')</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-up"></span> R output
</button>
<pre style="display:none"><code class="output r">## 
## gtRgnC> ## Obtain fullCov object
## gtRgnC> fullCov <- list('21'=genomeDataRaw$coverage)
## 
## gtRgnC> ## Assign chr lengths using hg19 information, use only first two regions
## gtRgnC> library('GenomicRanges')
</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## Loading required package: GenomeInfoDb
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-up"></span> R output
</button>
<pre style="display:none"><code class="output r">## 
## gtRgnC> data(hg19Ideogram, package = 'biovizBase', envir = environment())
## 
## gtRgnC> regions <- genomeRegions$regions[1:2]
## 
## gtRgnC> seqlengths(regions) <- seqlengths(hg19Ideogram)[names(seqlengths(regions))]
## 
## gtRgnC> ## Finally, get the region coverage
## gtRgnC> regionCov <- getRegionCoverage(fullCov=fullCov, regions=regions)
</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## 2014-09-30 14:50:45 getRegionCoverage: processing chr21
## 2014-09-30 14:50:45 getRegionCoverage: done processing chr21
</code></pre>
</div>
<p>Next, we load the coverage again using <code>which</code> but without any padding. We can see how the coverage is not the same by looking at the maximum coverage for each sample.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Illustrate reading data from a set of regions
test <- loadCoverage(files = files, chr = '21', cutoff = NULL, which = regions, protectWhich = 0)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## 2014-09-30 14:50:45 loadCoverage: finding chromosome lengths
## 2014-09-30 14:50:45 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam
## 2014-09-30 14:50:45 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam
## 2014-09-30 14:50:45 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009105_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009107_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009108_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009112_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009115_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009116_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009131_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009138_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009144_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009145_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009148_accepted_hits.bam
## 2014-09-30 14:50:46 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009151_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009152_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009153_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009159_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009161_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009163_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009164_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009167_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031812_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031835_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031867_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031868_accepted_hits.bam
## 2014-09-30 14:50:47 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031900_accepted_hits.bam
## 2014-09-30 14:50:48 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031904_accepted_hits.bam
## 2014-09-30 14:50:48 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031914_accepted_hits.bam
## 2014-09-30 14:50:48 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031936_accepted_hits.bam
## 2014-09-30 14:50:48 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031958_accepted_hits.bam
## 2014-09-30 14:50:48 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031960_accepted_hits.bam
## 2014-09-30 14:50:48 loadCoverage: applying the cutoff to the merged data
## 2014-09-30 14:50:48 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Some reads were ignored and thus the coverage is lower as can be seen below:
sapply(test$coverage, max) - sapply(genomeDataRaw$coverage, max)</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## ERR009101 ERR009102 ERR009105 ERR009107 ERR009108 ERR009112 ERR009115 
##         0         0         0         0        -1         0        -1 
## ERR009116 ERR009131 ERR009138 ERR009144 ERR009145 ERR009148 ERR009151 
##        -2        -2        -2        -1        -1        -3        -3 
## ERR009152 ERR009153 ERR009159 ERR009161 ERR009163 ERR009164 ERR009167 
##         0        -3        -3        -3        -1        -3        -3 
## SRR031812 SRR031835 SRR031867 SRR031868 SRR031900 SRR031904 SRR031914 
##         0        -1         0         0         0        -1         0 
## SRR031936 SRR031958 SRR031960 
##         0         0         0
</code></pre>
</div>
<p>When we re-load the data using some padding to the regions, we find that the coverage matches at all the bases.</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Illustrate reading data from a set of regions
test2 <- loadCoverage(files = files, chr = '21', cutoff = NULL, which = regions, protectWhich = 3e4)</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## 2014-09-30 14:50:49 loadCoverage: finding chromosome lengths
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009101_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009102_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009105_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009107_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009108_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009112_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009115_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009116_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009131_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009138_accepted_hits.bam
## 2014-09-30 14:50:49 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009144_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009145_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009148_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009151_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009152_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009153_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009159_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009161_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009163_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009164_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/ERR009167_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031812_accepted_hits.bam
## 2014-09-30 14:50:50 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031835_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031867_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031868_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031900_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031904_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031914_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031936_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031958_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: loading BAM file /Library/Frameworks/R.framework/Versions/3.1/Resources/library/derfinder/extdata/genomeData/SRR031960_accepted_hits.bam
## 2014-09-30 14:50:51 loadCoverage: applying the cutoff to the merged data
## 2014-09-30 14:50:51 filterData: originally there were 48129895 rows, now there are 48129895 rows. Meaning that 0 percent was filtered.
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Adding some padding to the regions helps get the same coverage
identical(sapply(test2$coverage, max), sapply(genomeDataRaw$coverage, max))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] TRUE
</code></pre>
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## A more detailed test reveals that the coverage matches at every base
all(mapply(function(x, y) { identical(x, y) }, test2$coverage, genomeDataRaw$coverage))</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] TRUE
</code></pre>
</div>
<p>How much padding you need to use will depend on your specific data set, and you might be comfortable getting approximately the same coverage values for the sake of greatly reducing the memory resources needed.</p>
<p>If you are under the case where you like to use a specific chromosome naming style but the raw data files use another one, you might need to use the <code>fileStyle</code> argument.</p>
<h2>Loading data in chunks</h2>
<p>If you prefer to do so, <code>fullCoverage()</code> and <code>loadCoverage()</code> can load the data of a chromosome in chunks using <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFiles.html">GenomicFiles</a>. This is controlled by whether you specify the <code>tilewidth</code> argument.</p>
<p>Notice that you might run into slight coverage errors near the borders of the tiles for the same reason that was illustrated previously when loading specific regions.</p>
<p>This approach is not necessarily more efficient and can be significantly time consuming if you use a small <code>tilewidth</code>.</p>
<h1>Flow charts</h1>
<h2>DER analysis flow chart</h2>
<p>The following figure illustrates how most of <code>derfinder</code>'s functions interact when performing a base-level differential expression analysis by calculating base-level F-statistics.</p>
<center>
<img src="http://lcolladotor.github.io/derfinder/fig/DERpathway.png" alt="DER flow" style="width: 500px;"/>
</center>
<p>Flow chart of the different processing steps (black boxes) that can be carried out using <code>derfinder</code> and the functions that perform these actions (in red). Input and output is shown in green boxes. Functions in blue are those applied to the results from multiple chromosomes (<code>mergeResults()</code> and <code>derfinderReport</code>). <code>regionReport</code> functions are shown in orange while <code>derfinderPlot</code> functions are shown in dark purple. Purple dotted arrow marks functions that require unfiltered base-level coverage.</p>
<h2><code>analyzeChr()</code> flow chart</h2>
<center>
<img src="http://lcolladotor.github.io/derfinder/fig/analyzeChr.png" alt="analyzeChr flow" style="width: 500px;"/>
</center>
<p>This figure shows in more detail the processing flow in <code>analyzeChr()</code>, which is the main function for identifying candidate differentially expressed regions (DERs) from the base-level F-statistics.</p>
<p>Many fine-tunning arguments can be passed to <code>analyzeChr()</code> to feed into the other functions. For example, you might want to use a smaller <code>chunksize</code> when pre-processing the coverage data (the default is 5 million): specially if you have hundreds of samples.</p>
<p>Another useful argument is <code>scalefac</code> (by default it's 32) which controls the scaling factor to use before the log2 transformation.</p>
<p>Furthermore, you might want to specify <code>maxClusterGap</code> to control the maximum gap between two regions before they are considered to be part of the same cluster.</p>
<h2><code>regionMatrix()</code> flow chart</h2>
<center>
<img src="http://lcolladotor.github.io/derfinder/fig/regionMatrix.png" alt="regionMatrix flow" style="width: 500px;"/>
</center>
<p>The above figure shows the functions used internally by <code>regionMatrix()</code> and processing steps. Overall, it is much simpler than <code>analyzeChr()</code>.</p>
<h1>Functions that use multiple cores</h1>
<p>Currently, the following functions can use multiple cores.</p>
<ul>
<li><code>calculatePvalues()</code>: 1 core per chunk of data to process.</li>
<li><code>calculateStats()</code>: 1 core per chunk of data to process.</li>
<li><code>coerceGR()</code>: 1 core per chromosome. This function is used by <code>createBw()</code>.</li>
<li><code>coverageToExon()</code>: 1 core per strand, then 1 core per chromosome.</li>
<li><code>loadCoverage()</code>: up to 1 core per tile when loading the data with <a href="http://www.bioconductor.org/packages/release/bioc/html/GenomicFiles.html">GenomicFiles</a>. Otherwise, no parallelization is used.</li>
<li><code>fullCoverage()</code>: 1 core per chromosome. In general, try to avoid using more than 10 cores as you might reach your maximum network speed and/or hard disk input/output seed. For the case described in <code>loadCoverage()</code>, you can specify how many cores to use per chromosome for the tiles using the <code>mc.cores.load</code> argument effectively resulting in <code>mc.cores</code> times <code>mc.cores.load</code> used (otherwise it's <code>mc.cores</code> squared).</li>
<li><code>getRegionCoverage()</code>: 1 core per chromosome.</li>
<li><code>regionMatrix()</code>: 1 core per chromosome.</li>
</ul>
<p>All parallel operations use <code>SnowParam()</code> from <a href="www.bioconductor.org/packages/release/bioc/html/BiocParallel.html">BiocParallel</a> when more than 1 core is being used. Otherwise, <code>SerialParam()</code> is used. Note that if you prefer to specify other types of parallelization you can do so by specifying the <code>BPPARAM</code> advanced argument.</p>
<p>Because <code>SnowParam()</code> requires <code>R</code> to load the necessary packages on each worker, the key function <code>fstats.apply()</code> was isolated in the <code>derfinderHelper</code> package. This package has much faster loading speeds than <code>derfinder</code> which greatly impacts performance on cases where the actual step of calculating the F-statistics is fast.</p>
<h1>Project organization</h1>
<p>For each project, we recommend the following organization.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|-derAnalysis
|---analysis01
|---analysis02</code></pre>
<p>That is, a main project directory with two initial directories. One for storing the coverage data, and one for storing each analysis: you might explore different models, cutoffs or other parameters.</p>
<p>You can then use <code>fullCoverage()</code>, save the result and also save the filtered coverage information for each chromosome separately. Doing so will result in the following structure.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---analysis01
|---analysis02</code></pre>
<p>Next, you can use <code>analyzeChr()</code> for each of the chromosomes of a specific analysis (say <em>analysis01</em>). Doing so will create several Rdata files per chromosome as shown below. <code>bash</code> scripts can be useful if you wish to submit one cluster job per chromosome. In general, you will use the same model and group information for each chromosome, so saving the information can be useful.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---analysis01
|-----models.Rdata
|-----groupInfo.Rdata
|-----chr1/
|-------chunksDir/
|-------logs/
|-------annotation.Rdata
|-------coveragePrep.Rdata
|-------fstats.Rdata
|-------optionsStats.Rdata
|-------regions.Rdata
|-------timeinfo.Rdata
|-----chr2/
...
|-----chrY/
|---analysis02</code></pre>
<p>Then use <code>mergeResults()</code> to pool together the results from all the chromosomes for a given analysis (here <em>analysis01</em>).</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---analysis01
|-----logs/
|-----fullAnnotatedRegions.Rdata
|-----fullFstats.Rdata
|-----fullNullSummary.Rdata
|-----fullRegions.Rdata
|-----fullTime.Rdata
|-----optionsMerge.Rdata
|-----chr1/
|-------chunksDir/
|-------logs/
|-------annotation.Rdata
|-------coveragePrep.Rdata
|-------fstats.Rdata
|-------optionsStats.Rdata
|-------regions.Rdata
|-------timeinfo.Rdata
|-----chr2/
...
|-----chrY/
|---analysis02</code></pre>
<p>Finally, you might want to use <code>derfinderReport()</code> from <code>regionReport</code> to create a HTML report of the results.</p>
<h1>bash scripts</h1>
<p>For interacting between <code>bash</code> and <code>R</code> we have found quite useful the <a href="http://cran.r-project.org/web/packages/getopt/index.html">getopt</a> package. Here we include an example <code>R</code> script that is controlled by a <code>bash</code> script which submits a job for each chromosome to analyze for a given analysis.</p>
<p>The two files, <em>derfinderAnalysis.R</em> and <em>derAnalysis.sh</em> should live under the <em>derAnalysis</em> directory.</p>
<pre><code>ProjectDir
|-derCoverageInfo
|---Chr1Cov.Rdata
|---Chr2Cov.Rdata
...
|---ChrYCov.Rdata
|---fullCov.Rdata
|-derAnalysis
|---derfinder-Analysis.R
|---derAnalysis.sh
|---analysis01
|---analysis02</code></pre>
<p>Then, you can simply use:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cd</span> /ProjectDir/derAnalysis
<span class="kw">sh</span> derAnalysis.sh analysis01</code></pre>
<p>to run <code>analyzeChr()</code> on all your chromosomes.</p>
<h2>derfinder-analysis.R</h2>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Run derfinder's analysis steps with timing info

## Load libraries
library("getopt")

## Available at https://github.com/lcolladotor/derfinder
library("derfinder")

## Specify parameters
spec <- matrix(c(
    'DFfile', 'd', 1, "character", "path to the .Rdata file with the results from loadCoverage()",
    'chr', 'c', 1, "character", "Chromosome under analysis. Use X instead of chrX.",
    'mcores', 'm', 1, "integer", "Number of cores",
    'verbose' , 'v', 2, "logical", "Print status updates",
    'help' , 'h', 0, "logical", "Display help"
), byrow=TRUE, ncol=5)
opt <- getopt(spec)

## Testing the script
test <- FALSE
if(test) {
    ## Speficy it using an interactive R session and testing
    test <- TRUE
}

## Test values
if(test){
    opt <- NULL
    opt$DFfile <- "/ProjectDir/derCoverageInfo/chr21Cov.Rdata"
    opt$chr <- "21"
    opt$mcores <- 1
    opt$verbose <- NULL
}

## if help was asked for print a friendly message
## and exit with a non-zero error code
if (!is.null(opt$help)) {
    cat(getopt(spec, usage=TRUE))
    q(status=1)
}

## Default value for verbose = TRUE
if (is.null(opt$verbose)) opt$verbose <- TRUE

if(opt$verbose) message("Loading Rdata file with the output from loadCoverage()")
load(opt$DFfile)

## Make it easy to use the name later. Here I'm assuming the names were generated using output='auto' in loadCoverage()
eval(parse(text=paste0("data <- ", "chr", opt$chr, "CovInfo")))
eval(parse(text=paste0("rm(chr", opt$chr, "CovInfo)")))

## Just for testing purposes
if(test) {
    tmp <- data
    tmp$coverage <- tmp$coverage[1:1e6, ]
    library("IRanges")
    tmp$position[which(tmp$pos)[1e6 + 1]:length(tmp$pos)] <- FALSE
    data <- tmp
}

## Load the models
load("models.Rdata")

## Load group information
load("groupInfo.Rdata")


## Run the analysis with lowMemDir
analyzeChr(chr=opt$chr, coverageInfo=data, models=models, cutoffFstat=1e-06, cutoffType="theoretical", nPermute=1000, seeds=seq_len(1000), maxClusterGap=3000, groupInfo=groupInfo, subject="hg19", mc.cores=opt$mcores, lowMemDir=paste0("chr", opt$chr, "/chunksDir"), verbose=opt$verbose, chunksize=1e5)

## Done
if(opt$verbose) {
    print(proc.time())
    print(sessionInfo(), locale=FALSE)
}</code></pre>
</div>
<p>Remember to modify the the script to fit your project.</p>
<h2>derAnalysis.sh</h2>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/sh</span>

<span class="co">## Usage</span>
<span class="co"># sh derAnalysis.sh analysis01</span>

<span class="co"># Directories</span>
<span class="ot">MAINDIR=</span>/ProjectDir
<span class="ot">WDIR=${MAINDIR}</span>/derAnalysis
<span class="ot">DATADIR=${MAINDIR}</span>/derCoverageInfo

<span class="co"># Define variables</span>
<span class="ot">SHORT=</span><span class="st">&#39;derA-01&#39;</span>
<span class="ot">PREFIX=$1</span>

<span class="co"># Construct shell files</span>
<span class="kw">for</span> <span class="kw">chrnum</span> in 22 21 Y 20 19 18 17 16 15 14 13 12 11 10 9 8 X 7 6 5 4 3 2 1
<span class="kw">do</span>
    <span class="kw">echo</span> <span class="st">&quot;Creating script for chromosome </span><span class="ot">${chrnum}</span><span class="st">&quot;</span>
    
    <span class="kw">if [[</span> <span class="ot">${chrnum}</span> <span class="ot">==</span> <span class="st">&quot;Y&quot;</span><span class="kw"> ]]</span>
    <span class="kw">then</span>
        <span class="ot">CORES=</span>2
    <span class="kw">else</span>
        <span class="ot">CORES=</span>6
    <span class="kw">fi</span>
    
    <span class="ot">chr=</span><span class="st">&quot;chr</span><span class="ot">${chrnum}</span><span class="st">&quot;</span>
    <span class="ot">outdir=</span><span class="st">&quot;</span><span class="ot">${PREFIX}</span><span class="st">/</span><span class="ot">${chr}</span><span class="st">&quot;</span>
    <span class="ot">sname=</span><span class="st">&quot;</span><span class="ot">${SHORT}</span><span class="st">.</span><span class="ot">${PREFIX}</span><span class="st">.</span><span class="ot">${chr}</span><span class="st">&quot;</span>
    <span class="kw">cat</span> <span class="kw">&gt;</span> <span class="ot">${WDIR}</span>/.<span class="ot">${sname}</span>.sh &lt;&lt;EOF
#!/bin/bash
#$ -cwd
#$ -m e
#$ -l mem_free=120G,h_vmem=60G,h_fsize=10G
#$ -N <span class="ot">${sname}</span>
#$ -pe local <span class="ot">${CORES}</span>

echo &quot;**** Job starts ****&quot;
date

# Create output directory 
mkdir -p <span class="ot">${WDIR}</span>/<span class="ot">${outdir}</span>
# Make logs directory
mkdir -p <span class="ot">${WDIR}</span>/<span class="ot">${outdir}</span>/logs

# run derfinder-analysis<span class="kw">.R</span>
cd <span class="ot">${WDIR}</span>/<span class="ot">${PREFIX}</span>/

# specific to our cluster
# see http://www.jhpce.jhu.edu/knowledge-base/environment-modules/
module load <span class="kw">R/3.1.x</span>
Rscript <span class="ot">${WDIR}</span>/derfinder2-analysis.R -d &quot;<span class="ot">${DATADIR}</span>/<span class="ot">${chr}</span>Cov.Rdata&quot; -c &quot;<span class="ot">${chrnum}</span>&quot; -m <span class="ot">${CORES}</span> -v TRUE

# Move log files into the logs directory
mv <span class="ot">${WDIR}</span>/<span class="ot">${sname}</span>.* <span class="ot">${WDIR}</span>/<span class="ot">${outdir}</span>/logs/

echo &quot;**** Job ends ****&quot;
date
EOF
    call=&quot;qsub .<span class="ot">${sname}</span>.sh&quot;
    <span class="ot">$call</span>
done</code></pre>
<p>Your cluster might specify memory requirements differently and you might need to use fewer or more cores depending on your data set.</p>
<h2>Full example</h2>
<p>A fully documented example is available at <a href="https://github.com/lcolladotor/derfinderExample">derfinderExample</a>.</p>
<p><strong>TODO</strong>: update the example to use <code>derfinder</code> version 0.99.0.</p>
<h1>Summary</h1>
<p>This vignette covered the most commonly used advanced arguments, details on how to load data, flow charts explaining the relationships between the functions, the recommended output organization, and example shell scripts for running the analysis.</p>
<h1>Origins</h1>
<p>This implementation of <code>derfinder</code> (<a href='https://github.com/lcolladotor/derfinder'>Collado-Torres, Frazee, Jaffe, and Leek, 2014</a>) has its origins in <a href="https://github.com/alyssafrazee/derfinder">Alyssa C. Frazee's derfinder</a> <a id='cite-originalder'></a>(<a href='#bib-originalder'>Frazee, Sabunciyan, Hansen, Irizzary, et al., 2013</a>). The statistical methods and implementation by now are very different.</p>
<h1>Citing <code>derfinder</code></h1>
<p>Please use:</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-down"></span> R source
</button>
<pre style=""><code class="source r">## Citation info
citation('derfinder')</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## 
## To cite package 'derfinder' in publications use:
## 
##   Leonardo Collado-Torres, Alyssa Frazee, Andrew Jaffe and Jeffrey
##   Leek (2014). derfinder: Fast differential expression analysis of
##   RNA-seq data at base-pair resolution. R package version 0.99.4.
##   https://github.com/lcolladotor/derfinder
## 
## A BibTeX entry for LaTeX users is
## 
##   @Manual{,
##     title = {derfinder: Fast differential expression analysis of RNA-seq data at base-pair
## resolution},
##     author = {Leonardo Collado-Torres and Alyssa Frazee and Andrew Jaffe and Jeffrey Leek},
##     year = {2014},
##     note = {R package version 0.99.4},
##     url = {https://github.com/lcolladotor/derfinder},
##   }
</code></pre>
</div>
<h1>Reproducibility</h1>
<p>This package was made possible thanks to:</p>
<ul>
<li>R <a id='cite-R'></a>(<a href='http://www.R-project.org/'>R Core Team, 2014</a>)</li>
<li>AnnotationDbi <a id='cite-AnnotationDbi'></a>(<a href='#bib-AnnotationDbi'>Pages, Carlson, Falcon, and Li, 2014</a>)</li>
<li>BiocParallel <a id='cite-BiocParallel'></a>(<a href='#bib-BiocParallel'>Morgan, Lang, and Thompson, 2014</a>)</li>
<li>bumphunter <a id='cite-bumphunter'></a>(<a href='#bib-bumphunter'>Irizarry, Aryee, Bravo, Hansen, et al., 2014</a>) and <a id='cite-bumphunterPaper'></a>(<a href='#bib-bumphunterPaper'>Jaffe, E, Murakami, Peter, et al., 2012</a>)</li>
<li>derfinderHelper <a id='cite-derfinderHelper'></a>(<a href='https://github.com/lcolladotor/derfinderHelper'>Collado-Torres, Jaffe, and Leek, 2014</a>)</li>
<li>GenomeInfoDb <a id='cite-GenomeInfoDb'></a>(<a href='#bib-GenomeInfoDb'>Arora, Morgan, Carlson, and Pages, 2014</a>)</li>
<li>GenomicAlignments <a id='cite-GenomicAlignments'></a>(<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pagès, Aboyoun, et al., 2013</a>)</li>
<li>GenomicFeatures (<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pagès, Aboyoun, et al., 2013</a>)</li>
<li>GenomicFiles <a id='cite-GenomicFiles'></a>(<a href='#bib-GenomicFiles'>Obenchain, Love, and Morgan, 2014</a>)</li>
<li>GenomicRanges (<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pagès, Aboyoun, et al., 2013</a>)</li>
<li>Hmisc <a id='cite-Hmisc'></a>(<a href='http://CRAN.R-project.org/package=Hmisc'>Jr, with contributions from Charles Dupont, and many others., 2014</a>)</li>
<li>IRanges (<a href='http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118'>Lawrence, Huber, Pagès, Aboyoun, et al., 2013</a>)</li>
<li>qvalue <a id='cite-qvalue'></a>(<a href='#bib-qvalue'>Dabney and Storey, 2014</a>)</li>
<li>Rsamtools <a id='cite-Rsamtools'></a>(<a href='http://bioconductor.org/packages/release/bioc/html/Rsamtools.html'>Morgan, Pagès, Obenchain, and Hayden, 2014</a>)</li>
<li>rtracklayer <a id='cite-rtracklayer'></a>(<a href='http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract'>Lawrence, Gentleman, and Carey, 2009</a>)</li>
<li>S4Vectors <a id='cite-S4Vectors'></a>(<a href='#bib-S4Vectors'>Pages, Lawrence, and Aboyoun, 2014</a>)</li>
<li>biovizBase <a id='cite-biovizBase'></a>(<a href='#bib-biovizBase'>Yin, Lawrence, and Cook, 2014</a>)</li>
<li>devtools <a id='cite-devtools'></a>(<a href='http://CRAN.R-project.org/package=devtools'>Wickham and Chang, 2014</a>)</li>
<li>ggplot2 <a id='cite-ggplot2'></a>(<a href='http://had.co.nz/ggplot2/book'>Wickham, 2009</a>)</li>
<li>knitcitations <a id='cite-knitcitations'></a>(<a href='https://github.com/cboettig/knitcitations'>Boettiger, 2014</a>)</li>
<li>knitr <a id='cite-knitr'></a>(<a href='http://www.crcpress.com/product/isbn/9781466561595'>Xie, 2014</a>)</li>
<li>knitrBootstrap <a id='cite-knitrBootstrap'></a>(<a href='https://github.com/jimhester/'>Hester, 2014</a>)</li>
<li>rmarkdown <a id='cite-rmarkdown'></a>(<a href='http://CRAN.R-project.org/package=rmarkdown'>Allaire, McPherson, Xie, Wickham, et al., 2014</a>)</li>
<li>testthat <a id='cite-testthat'></a>(<a href='http://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf'>Wickham, 2011</a>)</li>
<li>TxDb.Hsapiens.UCSC.hg19.knownGene <a id='cite-TxDbHsapiensUCSChg19knownGene'></a>(<a href='#bib-TxDbHsapiensUCSChg19knownGene'>Carlson, 2014</a>)</li>
</ul>
<p>Code for creating the vignette</p>
<div class="row">
<button class="source R toggle btn btn-xs btn-primary">
<span class="glyphicon glyphicon-chevron-up"></span> R source
</button>
<pre style="display:none"><code class="source r">## Create the vignette
library('knitrBootstrap') 

knitrBootstrapFlag <- packageVersion('knitrBootstrap') < '1.0.0'
if(knitrBootstrapFlag) {
    ## CRAN version
    library('knitrBootstrap')
    system.time(knit_bootstrap('derfinderAdvanced.Rmd', chooser=c('boot', 'code'), show_code = TRUE))
    unlink('derfinder.md')
} else {
    ## GitHub version
    library('rmarkdown')
    system.time(render('derfinderAdvanced.Rmd', 'knitrBootstrap::bootstrap_document'))
}
## Note: if you prefer the knitr version use:
# library('rmarkdown')
# system.time(render('derfinderAdvanced.Rmd', 'html_document'))
## Clean up
file.remove('derfinderAdvRef.bib')

## Extract the R code
library('knitr')
knit('derfinderAdvanced.Rmd', tangle = TRUE)</code></pre>
</div>
<p>Date the vignette was generated.</p>
<div class="row">
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## [1] "2014-09-30 14:50:52 EDT"
</code></pre>
</div>
<p>Wallclock time spent generating the vignette.</p>
<div class="row">
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">## Time difference of 21.95 secs
</code></pre>
</div>
<p><code>R</code> session information.</p>
<div class="row">
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## Session info--------------------------------------------------------------
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##  setting  value                       
##  version  R version 3.1.1 (2014-07-10)
##  system   x86_64, darwin10.8.0        
##  ui       AQUA                        
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  tz       America/New_York
</code></pre>
<button class="message R toggle btn btn-xs btn-info">
<span class="glyphicon glyphicon-chevron-up"></span> R message
</button>
<pre style="display:none"><code class="message r">## Packages------------------------------------------------------------------
</code></pre>
<button class="output R toggle btn btn-xs btn-success">
<span class="glyphicon glyphicon-chevron-down"></span> R output
</button>
<pre style=""><code class="output r">##  package           * version  date      
##  acepack             1.3.3.3  2013-05-03
##  AnnotationDbi       1.27.16  2014-09-26
##  base64enc           0.1.2    2014-06-26
##  BatchJobs           1.4      2014-09-24
##  BBmisc              1.7      2014-06-21
##  bibtex              0.3.6    2013-07-29
##  Biobase             2.25.0   2014-04-12
##  BiocGenerics      * 0.11.5   2014-09-13
##  BiocParallel        0.99.22  2014-09-24
##  biomaRt             2.21.1   2014-07-08
##  Biostrings          2.33.14  2014-09-09
##  bitops              1.0.6    2013-08-17
##  brew                1.0.6    2011-04-13
##  bumphunter          1.5.5    2014-08-29
##  checkmate           1.4      2014-09-03
##  cluster             1.15.3   2014-09-04
##  codetools           0.2.9    2014-08-21
##  DBI                 0.3.1    2014-09-24
##  derfinder         * 0.99.4   2014-09-30
##  derfinderHelper     0.99.2   2014-09-29
##  devtools          * 1.6      2014-09-23
##  digest              0.6.4    2013-12-03
##  doRNG               1.6      2014-03-07
##  evaluate            0.5.5    2014-04-29
##  fail                1.2      2013-09-19
##  foreach             1.4.2    2014-04-11
##  foreign             0.8.61   2014-03-28
##  formatR             1.0      2014-08-25
##  Formula             1.1.2    2014-07-13
##  futile.logger       1.3.7    2014-01-23
##  futile.options      1.0.0    2010-04-06
##  GenomeInfoDb      * 1.1.23   2014-09-28
##  GenomicAlignments   1.1.30   2014-09-23
##  GenomicFeatures     1.17.18  2014-09-29
##  GenomicFiles        1.1.19   2014-08-26
##  GenomicRanges     * 1.17.42  2014-09-23
##  Hmisc               3.14.5   2014-09-12
##  htmltools           0.2.6    2014-09-08
##  httr                0.5      2014-09-02
##  IRanges           * 1.99.28  2014-09-10
##  iterators           1.0.7    2014-04-11
##  knitcitations     * 1.0.2    2014-09-30
##  knitr               1.6      2014-05-25
##  knitrBootstrap    * 1.0.0    2014-09-30
##  lambda.r            1.1.6    2014-01-23
##  lattice             0.20.29  2014-04-04
##  latticeExtra        0.6.26   2013-08-15
##  locfit              1.5.9.1  2013-04-20
##  lubridate           1.3.3    2013-12-31
##  markdown            0.7.4    2014-08-24
##  Matrix              1.1.4    2014-06-15
##  matrixStats         0.10.0   2014-06-08
##  memoise             0.2.1    2014-04-22
##  nnet                7.3.8    2014-03-28
##  pkgmaker            0.22     2014-05-14
##  plyr                1.8.1    2014-02-26
##  qvalue              1.39.1   2014-05-27
##  R.methodsS3         1.6.1    2014-01-05
##  RColorBrewer        1.0.5    2011-06-17
##  Rcpp                0.11.2   2014-06-08
##  RCurl               1.95.4.3 2014-07-29
##  RefManageR          0.8.34   2014-08-19
##  registry            0.2      2012-01-24
##  RJSONIO             1.3.0    2014-07-28
##  rmarkdown         * 0.3.3    2014-09-17
##  rngtools            1.2.4    2014-03-06
##  rpart               4.1.8    2014-03-28
##  Rsamtools           1.17.34  2014-09-20
##  RSQLite             0.11.4   2013-05-26
##  rstudioapi          0.1      2014-03-27
##  rtracklayer         1.25.16  2014-09-10
##  S4Vectors         * 0.2.4    2014-09-14
##  sendmailR           1.2.1    2014-09-21
##  stringr             0.6.2    2012-12-06
##  survival            2.37.7   2014-01-22
##  XML                 3.98.1.1 2013-06-20
##  xtable              1.7.4    2014-09-12
##  XVector             0.5.8    2014-09-07
##  yaml                2.1.13   2014-06-12
##  zlibbioc            1.11.1   2014-04-19
##  source                                      
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  Github (lcolladotor/derfinder@031bb0f)      
##  Github (lcolladotor/derfinderHelper@54a9721)
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  Bioconductor                                
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  Github (cboettig/knitcitations@f567713)     
##  CRAN (R 3.1.0)                              
##  Github (jimhester/knitrBootstrap@76c41f0)   
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  Bioconductor                                
##  Bioconductor                                
##  CRAN (R 3.1.1)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.0)                              
##  CRAN (R 3.1.1)                              
##  Bioconductor                                
##  CRAN (R 3.1.0)                              
##  Bioconductor
</code></pre>
</div>
<h1>Bibliography</h1>
<p>This vignette was generated using <code>knitrBootstrap</code> (<a href='https://github.com/jimhester/'>Hester, 2014</a>) with <code>knitr</code> (<a href='http://www.crcpress.com/product/isbn/9781466561595'>Xie, 2014</a>) and <code>rmarkdown</code> (<a href='http://CRAN.R-project.org/package=rmarkdown'>Allaire, McPherson, Xie, Wickham, et al., 2014</a>) running behind the scenes.</p>
<p>Citations made with <code>knitcitations</code> (<a href='https://github.com/cboettig/knitcitations'>Boettiger, 2014</a>).</p>
<div class="row">
<p>
<a id='bib-rmarkdown'></a><a href="#cite-rmarkdown">[1]</a><cite> J. Allaire, J. McPherson, Y. Xie, H. Wickham, et al. <EM>rmarkdown: Dynamic Documents for R</EM>. R package version 0.3.3. 2014. URL: <a href="http://CRAN.R-project.org/package=rmarkdown">http://CRAN.R-project.org/package=rmarkdown</a>.</cite>
</p>
<p>
<a id='bib-GenomeInfoDb'></a><a href="#cite-GenomeInfoDb">[2]</a><cite> S. Arora, M. Morgan, M. Carlson and H. Pages. <EM>GenomeInfoDb: Utilities for manipulating chromosome and other 'seqname' identifiers</EM>. R package version 1.1.23. 2014.</cite>
</p>
<p>
<a id='bib-knitcitations'></a><a href="#cite-knitcitations">[3]</a><cite> C. Boettiger. <EM>knitcitations: Citations for knitr markdown files</EM>. R package version 1.0.2. 2014. URL: <a href="https://github.com/cboettig/knitcitations">https://github.com/cboettig/knitcitations</a>.</cite>
</p>
<p>
<a id='bib-TxDbHsapiensUCSChg19knownGene'></a><a href="#cite-TxDbHsapiensUCSChg19knownGene">[4]</a><cite> M. Carlson. <EM>TxDb.Hsapiens.UCSC.hg19.knownGene: Annotation package for TxDb object(s)</EM>. R package version 3.0.0. 2014.</cite>
</p>
<p>
<a id='bib-derfinder'></a><a href="#cite-derfinder">[5]</a><cite> L. Collado-Torres, A. Frazee, A. Jaffe and J. Leek. <EM>derfinder: Fast differential expression analysis of RNA-seq data at base-pair resolution</EM>. R package version 0.99.4. 2014. URL: <a href="https://github.com/lcolladotor/derfinder">https://github.com/lcolladotor/derfinder</a>.</cite>
</p>
<p>
<a id='bib-derfinderHelper'></a><a href="#cite-derfinderHelper">[6]</a><cite> L. Collado-Torres, A. Jaffe and J. Leek. <EM>derfinderHelper: derfinder helper package</EM>. R package version 0.99.2. 2014. URL: <a href="https://github.com/lcolladotor/derfinderHelper">https://github.com/lcolladotor/derfinderHelper</a>.</cite>
</p>
<p>
<a id='bib-qvalue'></a><a href="#cite-qvalue">[7]</a><cite> A. Dabney and J. D. Storey. <EM>qvalue: Q-value estimation for false discovery rate control</EM>. R package version 1.39.1. 2014.</cite>
</p>
<p>
<a id='bib-originalder'></a><a href="#cite-originalder">[8]</a><cite> A. C. Frazee, S. Sabunciyan, K. D. Hansen, R. A. Irizzary, et al. “Differential expression analysis of RNA-seq data at single-base resolution”. In: <EM>Biostatistics</EM> (2013).</cite>
</p>
<p>
<a id='bib-knitrBootstrap'></a><a href="#cite-knitrBootstrap">[9]</a><cite> J. Hester. <EM>knitrBootstrap: Knitr Bootstrap framework.</EM> R package version 1.0.0. 2014. URL: <a href="https://github.com/jimhester/">https://github.com/jimhester/</a>.</cite>
</p>
<p>
<a id='bib-bumphunter'></a><a href="#cite-bumphunter">[10]</a><cite> R. A. Irizarry, M. Aryee, H. C. Bravo, K. D. Hansen, et al. <EM>bumphunter: Bump Hunter</EM>. R package version 1.5.5. 2014.</cite>
</p>
<p>
<a id='bib-bumphunterPaper'></a><a href="#cite-bumphunterPaper">[11]</a><cite> Jaffe, A. E, Murakami, Peter, et al. “Bump hunting to identify differentially methylated regions in epigenetic epidemiology studies”. In: <EM>International Journal of Epidemiology</EM> (2012).</cite>
</p>
<p>
<a id='bib-Hmisc'></a><a href="#cite-Hmisc">[12]</a><cite> F. E. H. Jr, with contributions from Charles Dupont and many others. <EM>Hmisc: Harrell Miscellaneous</EM>. R package version 3.14-5. 2014. URL: <a href="http://CRAN.R-project.org/package=Hmisc">http://CRAN.R-project.org/package=Hmisc</a>.</cite>
</p>
<p>
<a id='bib-rtracklayer'></a><a href="#cite-rtracklayer">[13]</a><cite> M. Lawrence, R. Gentleman and V. Carey. “rtracklayer: an R package for interfacing with genome browsers”. In: <EM>Bioinformatics</EM> 25 (2009), pp. 1841-1842. DOI: <a href="http://dx.doi.org/10.1093/bioinformatics/btp328">10.1093/bioinformatics/btp328</a>. URL: <a href="http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract">http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract</a>.</cite>
</p>
<p>
<a id='bib-GenomicAlignments'></a><a href="#cite-GenomicAlignments">[14]</a><cite> M. Lawrence, W. Huber, H. Pagès, P. Aboyoun, et al. “Software for Computing and Annotating Genomic Ranges”. In: <EM>PLoS Computational Biology</EM> 9 (8 2013). DOI: <a href="http://dx.doi.org/10.1371/journal.pcbi.1003118">10.1371/journal.pcbi.1003118</a>. URL: <a href="http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118}.">http://www.ploscompbiol.org/article/info%3Adoi%2F10.1371%2Fjournal.pcbi.1003118}.</a>
</p>
</cite>
</p>
<p>
<a id='bib-BiocParallel'></a><a href="#cite-BiocParallel">[15]</a><cite> M. Morgan, M. Lang and R. Thompson. <EM>BiocParallel: Bioconductor facilities for parallel evaluation</EM>. R package version 0.99.22. 2014.</cite>
</p>
<p>
<a id='bib-Rsamtools'></a><a href="#cite-Rsamtools">[16]</a><cite> M. Morgan, H. Pagès, V. Obenchain and N. Hayden. <EM>Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import</EM>. R package version 1.17.34. 2014. URL: <a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">http://bioconductor.org/packages/release/bioc/html/Rsamtools.html</a>.</cite>
</p>
<p>
<a id='bib-GenomicFiles'></a><a href="#cite-GenomicFiles">[17]</a><cite> V. Obenchain, M. Love and M. Morgan. <EM>GenomicFiles: Distributed computing by file or by range</EM>. R package version 1.1.19. 2014.</cite>
</p>
<p>
<a id='bib-AnnotationDbi'></a><a href="#cite-AnnotationDbi">[18]</a><cite> H. Pages, M. Carlson, S. Falcon and N. Li. <EM>AnnotationDbi: Annotation Database Interface</EM>. R package version 1.27.16. 2014.</cite>
</p>
<p>
<a id='bib-S4Vectors'></a><a href="#cite-S4Vectors">[19]</a><cite> H. Pages, M. Lawrence and P. Aboyoun. <EM>S4Vectors: S4 implementation of vectors and lists</EM>. R package version 0.2.4. 2014.</cite>
</p>
<p>
<a id='bib-R'></a><a href="#cite-R">[20]</a><cite> R Core Team. <EM>R: A Language and Environment for Statistical Computing</EM>. R Foundation for Statistical Computing. Vienna, Austria, 2014. URL: <a href="http://www.R-project.org/">http://www.R-project.org/</a>.</cite>
</p>
<p>
<a id='bib-ggplot2'></a><a href="#cite-ggplot2">[21]</a><cite> H. Wickham. <EM>ggplot2: elegant graphics for data analysis</EM>. Springer New York, 2009. ISBN: 978-0-387-98140-6. URL: <a href="http://had.co.nz/ggplot2/book">http://had.co.nz/ggplot2/book</a>.</cite>
</p>
<p>
<a id='bib-testthat'></a><a href="#cite-testthat">[22]</a><cite> H. Wickham. “testthat: Get Started with Testing”. In: <EM>The R Journal</EM> 3 (2011), pp. 5–10. URL: <a href="http://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">http://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>.</cite>
</p>
<p>
<a id='bib-devtools'></a><a href="#cite-devtools">[23]</a><cite> H. Wickham and W. Chang. <EM>devtools: Tools to make developing R code easier</EM>. R package version 1.6. 2014. URL: <a href="http://CRAN.R-project.org/package=devtools">http://CRAN.R-project.org/package=devtools</a>.</cite>
</p>
<p>
<a id='bib-knitr'></a><a href="#cite-knitr">[24]</a><cite> Y. Xie. “knitr: A Comprehensive Tool for Reproducible Research in R”. In: <EM>Implementing Reproducible Computational Research</EM>. Ed. by V. Stodden, F. Leisch and R. D. Peng. ISBN 978-1466561595. Chapman and Hall/CRC, 2014. URL: <a href="http://www.crcpress.com/product/isbn/9781466561595">http://www.crcpress.com/product/isbn/9781466561595</a>.</cite>
</p>
<p>
<a id='bib-biovizBase'></a><a href="#cite-biovizBase">[25]</a><cite> T. Yin, M. Lawrence and D. Cook. <EM>biovizBase: Basic graphic utilities for visualization of genomic data.</EM> R package version 1.13.11. 2014.</cite>
</p>
</div>
</div>
</div>
<div class="navbar navbar-fixed-bottom navbar-inverse">
<div class="container">
<div class="navbar-header">
<button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-responsive-collapse">
<span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span>
</button>
</div>
<div id="bottom-navbar" class="navbar-collapse collapse navbar-responsive-collapse">
<ul class="nav navbar-nav navbar-right">
<li class="nav">
<p class="navbar-text">
Toggle
</p>
</li>
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Code <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="dropdown-header">
Languages
</li>
<li class="active">
<a href="#" class="toggle-global source R" type="source.R">R</a>
</li>
<li>
<a href="#" type="all-source" class="toggle-global">All</a>
</li>
</ul>
</li>
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Output <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="dropdown-header">
Type
</li>
<li class="active">
<a href="#" class="toggle-global output" type="output">output</a>
</li>
<li class="active">
<a href="#" class="toggle-global message" type="message">message</a>
</li>
<li>
<a href="#" type="all-output" class="toggle-global">All</a>
</li>
</ul>
</li>
<li class="active">
<a href="#" type="figure" class="toggle-global">Figures</a>
</li>
</ul>
<ul class="nav navbar-nav navbar-left">
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Bootstrap theme <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="active">
<a class="theme-switch" title="Default" href="#">Default</a>
</li>
<li>
<a class="theme-switch" title="Amelia" href="#">Amelia</a>
</li>
<li>
<a class="theme-switch" title="Cerulean" href="#">Cerulean</a>
</li>
<li>
<a class="theme-switch" title="Cosmo" href="#">Cosmo</a>
</li>
<li>
<a class="theme-switch" title="Cyborg" href="#">Cyborg</a>
</li>
<li>
<a class="theme-switch" title="Journal" href="#">Journal</a>
</li>
<li>
<a class="theme-switch" title="Flatly" href="#">Flatly</a>
</li>
<li>
<a class="theme-switch" title="Readable" href="#">Readable</a>
</li>
<li>
<a class="theme-switch" title="Simplex" href="#">Simplex</a>
</li>
<li>
<a class="theme-switch" title="Slate" href="#">Slate</a>
</li>
<li>
<a class="theme-switch" title="Spacelab" href="#">Spacelab</a>
</li>
<li>
<a class="theme-switch" title="United" href="#">United</a>
</li>
</ul>
</li>
</ul>
<ul class="nav navbar-nav navbar-left">
<li class="dropup">
<a href="#" class="dropdown-toggle" data-toggle="dropdown">Code highlighting <b class="caret"></b></a>
<ul class="dropdown-menu">
<li class="active">
<a class="highlight-switch" title="HighlightJs" href="#">HighlightJs</a>
</li>
<li>
<a class="highlight-switch" title="Dark" href="#">Dark</a>
</li>
<li>
<a class="highlight-switch" title="Far" href="#">Far</a>
</li>
<li>
<a class="highlight-switch" title="Idea" href="#">Idea</a>
</li>
<li>
<a class="highlight-switch" title="Sunburst" href="#">Sunburst</a>
</li>
<li>
<a class="highlight-switch" title="Zenburn" href="#">Zenburn</a>
</li>
<li>
<a class="highlight-switch" title="Visual Studio" href="#">Visual Studio</a>
</li>
<li>
<a class="highlight-switch" title="Ascetic" href="#">Ascetic</a>
</li>
<li>
<a class="highlight-switch" title="Magula" href="#">Magula</a>
</li>
<li>
<a class="highlight-switch" title="Github" href="#">Github</a>
</li>
<li>
<a class="highlight-switch" title="Google Code" href="#">Google Code</a>
</li>
<li>
<a class="highlight-switch" title="Brown Paper" href="#">Brown Paper</a>
</li>
<li>
<a class="highlight-switch" title="School Book" href="#">School Book</a>
</li>
<li>
<a class="highlight-switch" title="IR Black" href="#">IR Black</a>
</li>
<li>
<a class="highlight-switch" title="Solarized - Dark" href="#">Solarized - Dark</a>
</li>
<li>
<a class="highlight-switch" title="Solarized - Light" href="#">Solarized - Light</a>
</li>
<li>
<a class="highlight-switch" title="Arta" href="#">Arta</a>
</li>
<li>
<a class="highlight-switch" title="Monokai" href="#">Monokai</a>
</li>
<li>
<a class="highlight-switch" title="Xcode" href="#">Xcode</a>
</li>
<li>
<a class="highlight-switch" title="Pojoaque" href="#">Pojoaque</a>
</li>
<li>
<a class="highlight-switch" title="Rainbow" href="#">Rainbow</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow" href="#">Tomorrow</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night" href="#">Tomorrow Night</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night Bright" href="#">Tomorrow Night Bright</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night Blue" href="#">Tomorrow Night Blue</a>
</li>
<li>
<a class="highlight-switch" title="Tomorrow Night Eighties" href="#">Tomorrow Night Eighties</a>
</li>
</ul>
</li>
</ul>
</div>
</div>
</div>
<link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" title="Default" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/amelia/bootstrap.min.css" title="Amelia" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/cerulean/bootstrap.min.css" title="Cerulean" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/cosmo/bootstrap.min.css" title="Cosmo" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/cyborg/bootstrap.min.css" title="Cyborg" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/journal/bootstrap.min.css" title="Journal" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/flatly/bootstrap.min.css" title="Flatly" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/readable/bootstrap.min.css" title="Readable" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/simplex/bootstrap.min.css" title="Simplex" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/slate/bootstrap.min.css" title="Slate" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/spacelab/bootstrap.min.css" title="Spacelab" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://netdna.bootstrapcdn.com/bootswatch/3.0.0/united/bootstrap.min.css" title="United" media="screen"></link><link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" title="HighlightJs" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/dark.min.css" title="Dark" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/far.min.css" title="Far" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/idea.min.css" title="Idea" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/sunburst.min.css" title="Sunburst" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/zenburn.min.css" title="Zenburn" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/vs.min.css" title="Visual Studio" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/ascetic.min.css" title="Ascetic" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/magula.min.css" title="Magula" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/github.min.css" title="Github" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/googlecode.min.css" title="Google Code" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/brown_paper.min.css" title="Brown Paper" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/school_book.min.css" title="School Book" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/ir_black.min.css" title="IR Black" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/solarized_dark.min.css" title="Solarized - Dark" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/solarized_light.min.css" title="Solarized - Light" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/arta.min.css" title="Arta" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/monokai.min.css" title="Monokai" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/xcode.min.css" title="Xcode" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/pojoaque.min.css" title="Pojoaque" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/rainbow.min.css" title="Rainbow" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow.min.css" title="Tomorrow" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night.min.css" title="Tomorrow Night" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night-bright.min.css" title="Tomorrow Night Bright" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night-blue.min.css" title="Tomorrow Night Blue" media="screen"></link> <link rel="alternate stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/tomorrow-night-eighties.min.css" title="Tomorrow Night Eighties" media="screen"></link>
</div>
<div id="push">

</div>
<div id="footer">
<div class="container">
<p class="text-muted" id="credit">
Styled with <a href="https://github.com/jimhester/knitrBootstrap">knitrBootstrap</a>
</p>
</div>
</div>
<link rel="stylesheet" id="theme" href="https://netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" media="screen"></link><link rel="stylesheet" id="highlight" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/7.3/styles/default.min.css" media="screen"></link>
</div>
</body>
</html>
